Initial React v0.14 definitions

jbrantly opened the pull request:
This is a ~~WIP~~ PR for React v0.14 definitions. This supercedes and incorporates feedback from #5828

**Things done:**
- [x] Created separate definition files for the various new packages like `react-dom` and `react-addons-*`
- [x] Created/updated definitions for `react-dom` and `react-dom/server` and removed those from `react`
  - [x] Moved and renamed `batchedUpdates` addon to `unstable_batchedUpdates`
  - [x] Added `unstable_renderSubtreeIntoContainer`
  - [x] Added `version`
- [x] Moved addon definitions to their respective individual files
- [x] Removed `react/addons` (yay!!)
- [x] Removed `initializeTouchEvents`
- [x] Removed `classSet`
- [x] Removed `setProps` and `replaceProps` since they're deprecated
- [x] Removed `getDOMNode` since it's deprecated
- [x] Removed `cloneWithProps` since it's deprecated
- [x] Refactored `react-global.d.ts` to pull in definitions from the various addon definitions files instead of duplicating
- [x] Added legacy v0.13 files
- [x] React.Children.toArray - done by @thasner 
- [x] React.Children.map - done by @thasner
- [x] Update tests
- [x] Remove `react-dom` folder based on discussion in #5828
- [x] New `ref` behavior for DOM elements - done by @vsiao 
- [x] Stateless function components - done by @vsiao 
- [x] shallowCompare addon - done by @pspeter3 
- [x] CSSTransitionGroup updates
- [x] New elements/attributes - done by @vsiao 
- [x] per @drbignik, "valueLink support should also be added to HTMLAttributes to support LinkedStateMixin" - done by @vsiao 

~~I've also not addressed updates to elements/attributes. I plan to make a thorough update to these at a later date.~~ @vsiao did this

**Questions:**
- ~~`cloneWithProps` addon is deprecated. Should we just remove it?~~ removed per discussion
- ~~I've removed other deprecated items already since TS doesn't really have a way to show that something is deprecated but it's useful to use TS when upgrading. Others might feel differently.~~
- ~~Should we make individual test files for each addon definition or keep a monolithic test file like today? I'm thinking monolithic but would like opinions. One thing to keep in mind is #6125.~~ seems like we've gone monolithic for now
- ~~Should we put the existing v0.13 definitions in a legacy folder?~~ added using DT conventions

If anyone wants collab on my fork to help out just ask.

Not sure who all is interested in this so I'm just going to ping everyone who has had passing interest in the past: @pspeter3 @vsiao @basarat @johnnyreilly @RyanCavanaugh @ngbrown @B-Stefan @use-strict

johnnyreilly commented:
Re: deprecated, I think remove - it's embracing the inevitable. If people shout we could look at adding back.

I favour breaking up the tests but I don't really mind. Monolithic is fine.

use-strict commented:
For now, I'm thinking you should move the old v0.13 definitions to a separate legacy folder instead of overwriting them.

I'm also for splitting the tests.

jbrantly commented:
> For now, I'm thinking you should move the old v0.13 definitions to a separate legacy folder instead of overwriting them.

I actually meant to put that under "Questions". We did that with v0.12, it makes sense to me as well.

> I favour breaking up the tests but I don't really mind
> I'm also for splitting the tests.

One "gotcha" with the tests is #6125. Essentially if we want to split and also test both the module and the global version, we'd wind up with N modules * 2 which here = 20 test files. If we split should we forego testing the global version? Or should we just suck it up and have 20 test files? Or some other solution that I'm (hopefully) missing.

use-strict commented:
@jbrantly , in this case it's probably overkill. All addons in a single file (x2) should be fine.

pspeter3 commented on react/react-dom.d.ts:
```diff
@@ -0,0 +1,71 @@
+
+declare namespace __React {
+    
+    module __DOM {
```
Why is this a `module` instead of a `namespace`?

jbrantly replied:
I think you're right, it should be namespace.

pspeter3 commented:
I think that we should remove the deprecated items as well. @jbrantly, what is the best way to help out?

jbrantly commented:
> what is the best way to help out?

I've given you collab access to my fork so you can push to this PR if you want. Feel free to take anything on the list, I haven't had a chance to do anything more yet.

pspeter3 commented:
Thanks!

jbrantly commented:
I just rebased on top of master to pull in #6236. I'm going to start working on some of the various items that were commented on.

jbrantly commented:
I've addressed the various comments in the thread so far:

- Removed `cloneWithProps` @johnnyreilly @pspeter3 
- Now exporting the type as well as the variable for `CSSTransitionGroup`, `TransitionGroup`, `PureRenderMixin` and `LinkedStateMixin`. @use-strict 
- Added legacy files for v0.13 @use-strict 
- Fix minor naming issue @pspeter3 

Plenty of stuff still left to do though.

johnnyreilly commented:
Good work!

use-strict commented:
:+1:

vvakame commented:
travis ci failed. please merge master into this branch.

jbrantly commented:
@vvakame This is still a WIP and the tests haven't been updated yet :smile: That's why travis is failing. I'll ping when everything is done and ready to be merged.

pspeter3 commented:
I'm curious about how we're going to do stateless function components. They don't seem like they would have an obvious type signature except for `<T extends React.Props<any>>(props: T) => React.ReactElement<any>`. Do people agree?

dallonf commented:
@pspeter3 According to this ticket, it doesn't look like function components are going to be supported by TypeScript yet: https://github.com/Microsoft/TypeScript/issues/4861

jbrantly commented:
@dallonf Thanks for that link. I've been meaning to open that issue for the longest time but never got around to it. I'm glad to see someone else already did it.

@pspeter3 That seems right to me. I think the main thing will be updating `createElement` to accept them as component types.

RyanCavanaugh commented:
@dallonf FWIW I intend to address Microsoft/TypeScript#4861 this week

jbrantly commented:
@RyanCavanaugh that is excellent news! Looking forward to it.

drbignik commented:
valueLink support should also be added to HTMLAttributes to support LinkedStateMixin.

alexgorbatchev commented on react/react-dom.d.ts:
```diff
@@ -0,0 +1,71 @@
+// Type definitions for React v0.14 (react-dom)
```
I think `react-dom` is a completely separate package from React which is already published here https://github.com/borisyankov/DefinitelyTyped/blob/master/react-dom/react-dom.d.ts 

I also suggest changing the name to `react-dom` instead what's here now. Basically I would recommend splitting out all of the modules into separate packages to match NPM.

jbrantly replied:
See #5828 for a discussion on this.

vsiao commented:
Sorry, I'm a little late to the party. Anything I can help with @jbrantly?

jbrantly commented:
@vsiao Better late than never! Glad you'll be able to take a look at these before they're merged. I've made you a collaborator so if you want to push feel free. I'm about 80% done with fixing the tests, but anything else on the list is up for grabs. I believe I'll be getting some time to try to finish this up in the next couple days so if you do start working on something maybe call it here to avoid duplicate effort.

pspeter3 commented:
I can take the `shallowCompare` add-on and a crack at Stateless function components.

jbrantly commented:
I'm calling CSSTransitionGroup after I finish the tests.

thasner commented:
@jbrantly I can write typings for `React.Children.toArray` and `React.Children.map`

jbrantly commented:
@thasner I gave you collab and updated the OP. Thanks!

thasner commented:
@jbrantly I could also fix the currently failing tests. Would that conflict with anything you've done so far?

jbrantly commented:
@thasner Yea that'd conflict. I'm like 90% done with them. I ran into some issues over the weekend when I was trying to finish them up. I got through the issues but still need to wrap up that last 10%.

johnnyreilly commented:
Hi all,

You may have noticed that https://github.com/borisyankov/DefinitelyTyped has now 301'd to https://github.com/DefinitelyTyped/DefinitelyTyped.  Hopefully this should be seemless for the work you're doing here; do let us know if there are any problems.

vsiao commented:
I'm working on the new DOM refs but finding it difficult because I can't tell which tests are broken because of my changes. Fixing the currently failing tests would be much appreciated. Let me know if you want any help!

jbrantly commented:
@vsiao @thasner Sorry for the delay, but the tests are now done. The snag that I hit was that the addons were not really exposing related interfaces in the module version (for example, `CSSTransitionGroup` did not expose `CSSTransitionGroupProps`). I originally tried exporting each module as both a value and a namespace with all of the interfaces. I had limited success (eventually) but ultimately decided against that approach. Instead interfaces like `CSSTransitionGroupProps` are simply appended to the `React` namespace. This is actually exactly how it worked in v0.13. I was also able to clean up the `react-global.ts` file quite a bit by adding the variable exports to the individual addon files.

jbrantly commented:
I can't update dependent tests (in particular, `fixed-data-table`) without rebasing from master. Normally this wouldn't be an issue but since others are working on this feature branch I don't want to `--force` push behind their backs. So I think I'll wait until everything else is done and I'll update dependent tests last.

vvakame commented:
`git merge --no-ff master` ?

pspeter3 commented:
I think it is safe to rebase / I would be willing to fix any merge conflicts I will have.

thasner commented:
no worries @jbrantly , thanks for coordinating all of this!

:+1: for `git merge --no-ff master`

jamiewinder commented:
Just noticed an issue in both the current react.d.ts and this one - the `value` and `defaultValue` properties are defined as `string` types, but these properties also accept `Array<string>` in the case of `<select multiple ... />` elements.

pspeter3 commented:
Added shallow-compare

vsiao commented:
Added types for stateless components (stole the task from @pspeter3) and merged with master.

vsiao commented:
@jbrantly looks like we're done with the remaining tasks except for CSSTransitionGroup updates and testing. I've already been adding tests as I go. Do you know what still needs to be tested?

jbrantly commented:
@vsiao Awesome! I just finished up the CSSTransitionGroup update and also fixing dependency tests. My "Test, test, test" comment was more about upgrading real-world applications and not so much the unit tests, but I'm not sure if that's practical. I've removed that item.

There is only one remaining item I think related to the `valueLink` changes and the material-ui definitions (see my comment on 813f3b0).

vsiao commented:
Done! Passing tests finally :smile:

jbrantly commented:
OK! I think everything is done and ready to be merged!

@vvakame

johnnyreilly commented:
Merged - well done!

pspeter3 commented:
Thanks for driving this @jbrantly

myitcv commented:
Thanks to everyone involved. Very much appreciated.

jbrantly commented:
Yup, thanks to everyone who contributed with code or comments!

jfahrenkrug commented:
Thank all of you so much for all your hard work on this!

adidahiya commented:
thanks all :star2: will start testing these definitions soon!

alexgorbatchev commented:
Amazing work, thank you so much everyone!

rhysd commented:
Thank you for the great work!

pserwylo commented:
Thanks all for your hard work on this. We've been using these while they have been in progress, and so far no problems at all.

twhitbeck commented:
I have an issue using ES6 module syntax in TypeScript/Visual Studio 2013.

```typescript
import * as update from 'react-addons-update';
```

error: "Module react-addons-update resolves to a non-module entity and cannot be imported using this construct"

```typescript
import update = require('react-addons-update');
```

No error. Do I have to use this syntax? I'm using es6 syntax everywhere else. :(

use-strict commented:
tl;dr: you have to use `import ... = require(...)` in this case.

`import ... from ...` is used for importing ES6 modules. `import * as ...` usually works with most commonjs because it translates to using the `exports` object. However, there is no equivalent to `module.exports = ` construct in the ES6 spec. Even the default export ends up as an `export.default = ...` and you cannot override the exports object.

Strate commented:
`type ReactInstance = Component<any, any> | Element` is wrong. Trying to accessing any `Element`'s property gives me a compilation error:
```
this.refs['refname'].scrollHeight // error TS2339: Property 'scrollHeight' does not exist on type 'Component<any, any> | Element'.
```
This happening because of [typescript union type spec](https://github.com/Microsoft/TypeScript/issues/805):

> The type A|B has a property P of type X|Y if A has a property P of type X and B has a property P of type Y.
*(from Properties section)*

Instead of using union type, there should be [intersection type](https://github.com/Microsoft/TypeScript/pull/3622), which completely fits to this case:

> The type A & B has a property P if A has a property P or B has a property P.
*(from Properties section)*

So, there should be `type ReactInstance =  Component<any, any> & Element`
